File: katas/python/OOP/inheritance/inheritance_bad.py
# Violación de Herencia: Código duplicado sin reutilización
# ❌ Problema: Lógica repetida en cada clase


class Dog:
    # ❌ Código duplicado en todas las clases de animales
    def __init__(self, name: str, age: int, weight: float):
        self.name = name
        self.age = age
        self.weight = weight

    # ❌ Métodos comunes duplicados
    def eat(self) -> None:
        print(f"{self.name} está comiendo")
        self.weight += 0.5

    def sleep(self) -> None:
        print(f"{self.name} está durmiendo")

    def get_info(self) -> str:
        return f"Nombre: {self.name}, Edad: {self.age} años, Peso: {self.weight}kg"

    # Método específico
    def bark(self) -> None:
        print(f"{self.name} dice: ¡Guau guau!")


class Cat:
    # ❌ Mismo código duplicado otra vez
    def __init__(self, name: str, age: int, weight: float):
        self.name = name
        self.age = age
        self.weight = weight

    # ❌ Métodos comunes duplicados exactamente igual
    def eat(self) -> None:
        print(f"{self.name} está comiendo")
        self.weight += 0.5

    def sleep(self) -> None:
        print(f"{self.name} está durmiendo")

    def get_info(self) -> str:
        return f"Nombre: {self.name}, Edad: {self.age} años, Peso: {self.weight}kg"

    # Método específico
    def meow(self) -> None:
        print(f"{self.name} dice: ¡Miau miau!")


class Bird:
    # ❌ Y otra vez el mismo código duplicado
    def __init__(self, name: str, age: int, weight: float):
        self.name = name
        self.age = age
        self.weight = weight

    # ❌ Métodos comunes duplicados una tercera vez
    def eat(self) -> None:
        print(f"{self.name} está comiendo")
        self.weight += 0.5

    def sleep(self) -> None:
        print(f"{self.name} está durmiendo")

    def get_info(self) -> str:
        return f"Nombre: {self.name}, Edad: {self.age} años, Peso: {self.weight}kg"

    # Método específico
    def chirp(self) -> None:
        print(f"{self.name} dice: ¡Pío pío!")


if __name__ == "__main__":
    # ❌ Mucho código duplicado y difícil de mantener
    print("=== Violación de Herencia ===")

    dog = Dog("Rex", 5, 25)
    cat = Cat("Luna", 3, 4)
    bird = Bird("Piolín", 1, 0.5)

    print(dog.get_info())
    dog.eat()
    dog.bark()

    print(cat.get_info())
    cat.eat()
    cat.meow()

    print(bird.get_info())
    bird.eat()
    bird.chirp()

    # ❌ Si necesito cambiar eat() o sleep(), debo modificar 3 clases
    # ❌ Si agrego más animales, más duplicación
    # ❌ Propenso a inconsistencias

File: katas/csharp/OOP/inheritance/inheritance-bad.cs
using System;

// Violación de Herencia: Código duplicado sin reutilización
// ❌ Problema: Lógica repetida en cada clase

public class Dog
{
  // ❌ Código duplicado en todas las clases de animales
  private string name;
  private int age;
  private double weight;

  public Dog(string name, int age, double weight)
  {
    this.name = name;
    this.age = age;
    this.weight = weight;
  }

  // ❌ Métodos comunes duplicados
  public void Eat()
  {
    Console.WriteLine($"{name} está comiendo");
    weight += 0.5;
  }

  public void Sleep()
  {
    Console.WriteLine($"{name} está durmiendo");
  }

  public string GetInfo()
  {
    return $"Nombre: {name}, Edad: {age} años, Peso: {weight}kg";
  }

  // Método específico
  public void Bark()
  {
    Console.WriteLine($"{name} dice: ¡Guau guau!");
  }
}

public class Cat
{
  // ❌ Mismo código duplicado otra vez
  private string name;
  private int age;
  private double weight;

  public Cat(string name, int age, double weight)
  {
    this.name = name;
    this.age = age;
    this.weight = weight;
  }

  // ❌ Métodos comunes duplicados exactamente igual
  public void Eat()
  {
    Console.WriteLine($"{name} está comiendo");
    weight += 0.5;
  }

  public void Sleep()
  {
    Console.WriteLine($"{name} está durmiendo");
  }

  public string GetInfo()
  {
    return $"Nombre: {name}, Edad: {age} años, Peso: {weight}kg";
  }

  // Método específico
  public void Meow()
  {
    Console.WriteLine($"{name} dice: ¡Miau miau!");
  }
}

public class Bird
{
  // ❌ Y otra vez el mismo código duplicado
  private string name;
  private int age;
  private double weight;

  public Bird(string name, int age, double weight)
  {
    this.name = name;
    this.age = age;
    this.weight = weight;
  }

  // ❌ Métodos comunes duplicados una tercera vez
  public void Eat()
  {
    Console.WriteLine($"{name} está comiendo");
    weight += 0.5;
  }

  public void Sleep()
  {
    Console.WriteLine($"{name} está durmiendo");
  }

  public string GetInfo()
  {
    return $"Nombre: {name}, Edad: {age} años, Peso: {weight}kg";
  }

  // Método específico
  public void Chirp()
  {
    Console.WriteLine($"{name} dice: ¡Pío pío!");
  }
}

// ❌ Mucho código duplicado y difícil de mantener
class Program
{
  static void Main()
  {
    Console.WriteLine("=== Violación de Herencia ===");

    var dog = new Dog("Rex", 5, 25);
    var cat = new Cat("Luna", 3, 4);
    var bird = new Bird("Piolín", 1, 0.5);

    Console.WriteLine(dog.GetInfo());
    dog.Eat();
    dog.Bark();

    Console.WriteLine(cat.GetInfo());
    cat.Eat();
    cat.Meow();

    Console.WriteLine(bird.GetInfo());
    bird.Eat();
    bird.Chirp();

    // ❌ Si necesito cambiar Eat() o Sleep(), debo modificar 3 clases
    // ❌ Si agrego más animales, más duplicación
    // ❌ Propenso a inconsistencias
  }
}
